src/Terminal.hs

```diff
@@ -0,0 +1,100 @@
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Terminal (terminalGame) where
+
+import Control.Applicative (liftA2)
+import Control.Monad ((<=<))
+import Data.Either.Extra (maybeToEither)
+import Data.Function (on)
+import Data.Map.Strict (empty, findWithDefault)
+import Data.Text qualified as Text
+import Data.Text.IO qualified as TextIO
+import GameLogic (Board (..), Coordinates, Player (..), nextTurn, playAtSpot, winner)
+import System.Console.ANSI (Color (..), ColorIntensity (..), ConsoleLayer (..), SGR (..), clearScreen, setCursorPosition, setSGR)
+import Text.Read (readMaybe)
+
+showPlayerT :: Player -> Text.Text
+showPlayerT = Text.pack . show
+
+showBoard :: Board -> Text.Text
+showBoard (Board board) =
+  "  0   1   2 \n"
+    <> Text.intercalate "\n ---+---+---\n" rows
+  where
+    rows = rowString <$> (zip [0 ..] $ replicate 3 [0 .. 2])
+    rowString :: (Int, [Int]) -> Text.Text
+    rowString (colIndex, row) =
+      (Text.pack . show) colIndex
+        <> " "
+        <> Text.intercalate " | " ((\t -> findWithDefault " " (t, colIndex) textBoard) <$> row)
+
+    textBoard = playerToSpotText <$> board
+
+    playerToSpotText :: Player -> Text.Text
+    playerToSpotText Circle = "0"
+    playerToSpotText Cross = "x"
+
+getCoordsAndPlayTurn :: Board -> Player -> IO (Board, Coordinates)
+getCoordsAndPlayTurn board turn = do
+  TextIO.putStrLn $ showPlayerT turn <> " turn"
+  coord <- getInputAndValidate parseIntTuple "Please Enter space seperated coordinates: x y"
+  case playAtSpot coord turn board of
+    Nothing -> do
+      setSGR [SetColor Foreground Dull Red]
+      TextIO.putStrLn "Invalid placement"
+      setSGR [Reset]
+      getCoordsAndPlayTurn board turn
+    Just newBoard -> pure (newBoard, coord)
+
+type Parser a = (Text.Text -> Either Text.Text a)
+
+getInputAndValidate :: Parser a -> Text.Text -> IO a
+getInputAndValidate validator message = do
+  TextIO.putStrLn message
+  input <- TextIO.getLine
+  case validator input of
+    Right value -> pure value
+    Left errorMessage -> do
+      setSGR [SetColor Foreground Dull Red]
+      TextIO.putStrLn errorMessage
+      setSGR [Reset]
+      getInputAndValidate validator message
+
+parseNumber :: Parser Int
+parseNumber = inBounds <=< maybeToEither "Input has to be a number" . readMaybe . Text.unpack
+  where
+    inBounds :: Int -> Either Text.Text Int
+    inBounds x
+      | x < 3 && x >= 0 = Right x
+      | otherwise = Left "Coordinate out of bounds"
+
+parseIntTuple :: Parser Coordinates
```

Comment:
It's not just an `Int` tuple, it's `Coordinates`. We want names to reflect the domain of the problem we're solving ("business logic"), not technicalities. Even more so with such a highly expressive static type system.